File: test/gleeunit/internal/gleeunit_gleam_panic_ffi.mjs
import { Ok, Error, Empty, NonEmpty } from "../../gleam.mjs";
import {
  GleamPanic,
  Todo,
  Panic,
  LetAssert,
  Assert,
  BinaryOperator,
  FunctionCall,
  OtherExpression,
  AssertedExpression,
  Literal,
  Expression,
  Unevaluated,
} from "./gleam_panic.mjs";

export function from_dynamic(error) {
  if (!(error instanceof globalThis.Error) || !error.gleam_error) {
    return new Error(undefined);
  }

  if (error.gleam_error === "todo") {
    return wrap(error, new Todo());
  }

  if (error.gleam_error === "panic") {
    return wrap(error, new Panic());
  }

  if (error.gleam_error === "let_assert") {
    let kind = new LetAssert(
      error.start,
      error.end,
      error.pattern_start,
      error.pattern_end,
      error.value,
    );
    return wrap(error, kind);
  }

  if (error.gleam_error === "assert") {
    let kind = new Assert(
      error.start,
      error.end,
      error.expression_start,
      assert_kind(error),
    );
    return wrap(error, kind);
  }

  return new Error(undefined);
}

function assert_kind(error) {
  if (error.kind == "binary_operator") {
    return new BinaryOperator(
      error.operator,
      expression(error.left),
      expression(error.right),
    );
  }

  if (error.kind == "function_call") {
    let list = new Empty();
    let i = error.arguments.length;
    while (i--) {
      list = new NonEmpty(expression(error.arguments[i]), list);
    }
    return new FunctionCall(list);
  }

  return new OtherExpression(expression(error.expression));
}

function expression(data) {
  const expression = new AssertedExpression(data.start, data.end, undefined);
  if (data.kind == "literal") {
    expression.kind = new Literal(data.value);
  } else if (data.kind == "expression") {
    expression.kind = new Expression(data.value);
  } else {
    expression.kind = new Unevaluated();
  }
  return expression;
}

function wrap(e, kind) {
  return new Ok(
    new GleamPanic(e.message, e.file, e.module, e.function, e.line, kind),
  );
}

File: test/gleeunit/internal/gleeunit_gleam_panic_ffi.erl
-module(gleeunit_gleam_panic_ffi).
-export([from_dynamic/1]).

from_dynamic(#{
  gleam_error := assert,
  start := Start,
  'end' := End,
  expression_start := EStart
} = E) ->
    wrap(E, {assert, Start, End, EStart, assert_kind(E)});
from_dynamic(#{
  gleam_error := let_assert,
  start := Start,
  'end' := End,
  pattern_start := PStart,
  pattern_end := PEnd,
  value := Value
} = E) ->
    wrap(E, {let_assert, Start, End, PStart, PEnd, Value});
from_dynamic(#{gleam_error := panic} = E) ->
    wrap(E, panic);
from_dynamic(#{gleam_error := todo} = E) ->
    wrap(E, todo);
from_dynamic(_) ->
    {error, nil}.

assert_kind(#{kind := binary_operator, left := L, right := R, operator := O}) ->
    {binary_operator, atom_to_binary(O), expression(L), expression(R)};
assert_kind(#{kind := function_call, arguments := Arguments}) ->
    {function_call, lists:map(fun expression/1, Arguments)};
assert_kind(#{kind := expression, expression := Expression}) ->
    {other_expression, expression(Expression)}.

expression(#{start := S, 'end' := E, kind := literal, value := Value}) ->
    {asserted_expression, S, E, {literal, Value}};
expression(#{start := S, 'end' := E, kind := expression, value := Value}) ->
    {asserted_expression, S, E, {expression, Value}};
expression(#{start := S, 'end' := E, kind := unevaluated}) ->
    {asserted_expression, S, E, unevaluated}.

wrap(#{
    gleam_error := _,
    file := File,
    message := Message,
    module := Module,
    function := Function,
    line := Line
}, Kind) ->
    {ok, {gleam_panic, Message, File, Module, Function, Line, Kind}}.
